torch/nn/parallel/scatter_gather.py

```diff
@@ -46,5 +46,8 @@ def gather_map(outputs):
             return Gather.apply(target_device, dim, *outputs)
         if out is None:
             return None
-        return type(out)(map(gather_map, zip(*outputs)))
+        if isinstance(out, dict):
+            return dict(zip(out.keys(),
+                            map(gather_map, zip(*[d.values() for d in outputs]))))
```

Comment:
Correct me if I'm wrong, but I don't think that `keys()` is necessarily ordered in the same way as `values()`. Could you do instead 
```python
keys = tuple(out.keys())
return dict([(k, gather_map([d[k] for d in outputs])) for k in keys])
```

Reply:
or better yet, iterate over `.items()`

Reply:
@apaszke Not necessarily. It is not guaranteed that every dict's `.items()` is of the same order. So it's still better to index into every dict using a reference dict's keys.

Reply:
Ah I haven't noticed that these are two different dicts. You're right.